code/modules/projectiles/projectile_parent.dm

```diff
@@ -756,6 +825,51 @@ ABSTRACT_TYPE(/datum/projectile)
 			src.ie_type = P.ie_type
 			src.impact_image_state = P.impact_image_state
 
+		// Spawn some particles if we hit something solid that isnt a human or a silicon
+		spawn_impact_particles(atom/hit, var/obj/projectile/O, x, y)
+			if (src.has_impact_particles && !ismob(hit))
+				//If we are underwater, we want bubbles and not sparks or smoke!
+				var/underwater = FALSE
+				if (istype(get_turf(O), /turf/space/fluid)) underwater = TRUE
+				else
+					var/turf/T = get_turf(O)
+					if (T?.active_liquid)
+						if(T.active_liquid.last_depth_level > 3)
+							underwater = TRUE
+				if (src.kinetic_impact)
+					var/new_impact_icon = hit.impact_icon
+					var/new_impact_icon_state = hit.impact_icon_state
+					//Bullet impacts create dust of the color of the hit thing
+					var/avrg_color = hit.get_average_color()
+					new /obj/effects/gunshot_impact/dust(get_turf(hit), x, y, -O.xo, -O.yo, damage, avrg_color, new_impact_icon, new_impact_icon_state)
+					if (underwater)
+						new /obj/effects/gunshot_impact/bubble(get_turf(hit), x, y, -O.xo, -O.yo, damage)
+					else
+						new /obj/effects/gunshot_impact/sparks(get_turf(hit), x, y, -O.xo, -O.yo, damage)
+						new /obj/effects/gunshot_impact/smoke(get_turf(hit), x, y, -O.xo, -O.yo, damage)
+				else
+					//Energy impacts create sparks of the color of the projectile
```

Comment:
Maybe try putting these on a plane above lighting? It looks odd for them to be hidden by darkness since they're meant to be some kind of energy splash. (might look weirder if they're fullbright but worth a try)